libkrun/libkruntime.c

```diff
@@ -342,126 +419,213 @@ libkruntime_done(void)
 #endif
 }
 
-uint64_t
-read_aperf(void)
+static uint64_t
+krun_read_aperf(int core)
 {
 #if defined(__linux__) && !defined(TRAVIS)
-    return sum_msr_all_cores(IA32_APERF, IA32_APERF_MASK);
+    return krun_read_msr(core, IA32_APERF) & IA32_APERF_MASK;
 #elif defined(__linux__) && defined(TRAVIS)
     return 0;
 #else
 #error "Unsupported platform"
 #endif
 }
 
-uint64_t
-read_mperf(void)
+static uint64_t
+krun_read_mperf(int core)
 {
 #if defined(__linux__) && !defined(TRAVIS)
-    return sum_msr_all_cores(IA32_MPERF, IA32_MPERF_MASK);
+    return krun_read_msr(core, IA32_MPERF) & IA32_MPERF_MASK;
 #elif defined(__linux__) && defined(TRAVIS)
     return 0;
 #else
 #error "Unsupported platform"
 #endif
 }
 
-#if defined(__linux__) && !defined(TRAVIS)
 /*
- * Sum the values of an MSR across all cores.
+ * Check for double precision loss.
  *
- * Arguments:
- *   msr_addr: address of the MSR to measure.
- *   mask: value to AND each core's MSR value with prior to summation.
+ * Since some languages cannot represent a uint64_t, we sometimes have to pass
+ * around a double. This is annoying since precision could be silently lost.
+ * This function makes loss of precision explicit, stopping the VM.
  *
- * In the future, a 'shift' argument may be necessary, should we ever read a
- * value from an MSR which doesn't sit flush with bit 0.
+ * We don't expect to actually see a crash since the TSR is zeroed at reboot,
+ * and our benchmarks are not long enough running to generate a large enough
+ * TSR value to cause precision loss (you would need a integer that would not
+ * fit in a 52-bit unsigned int before precision starts being lost in
+ * lower-order bits). Nevertheless, we check it.
+ *
+ * This routine comes at a small cost (a handful of asm instrs). Note that this
+ * cost is a drop in the ocean compared to benchmark workloads.
  */
-uint64_t
-sum_msr_all_cores(long msr_addr, uint64_t mask)
+double
+krun_u64_to_double(uint64_t u64_val)
+{
+    double d_val = (double) u64_val;
+    uint64_t u64_val2 = (uint64_t) d_val;
+
+    if (u64_val != u64_val2) {
+        fprintf(stderr, "Loss of precision detected!\n");
+        fprintf(stderr, "%" PRIu64 " != %" PRIu64 "\n", u64_val, u64_val2);
+        exit(EXIT_FAILURE);
+    }
+    // (otherwise all is well)
+    return d_val;
+}
+
+/*
+ * Load up the results struct.
+ */
+void
+krun_measure(int mdata_idx)
 {
-    int core;
-    uint64_t sum = 0, val;
+    struct krun_data *data = &(krun_mdata[mdata_idx]);
 
-    for (core = 0; core < num_msr_nodes; core++) {
-        val = read_msr(core, msr_addr) & mask;
+    KDATA_BOUNDS_CHECK(mdata_idx);
 
-        // Check for overflow
-        if (sum > UINT64_MAX - val) {
-            fprintf(stderr, "Sum overflow!: "
-                    "%" PRIu64 " + %" PRIu64 "\n", sum, val);
-            exit(EXIT_FAILURE);
+#if defined(__linux__) && !defined(TRAVIS)
+    // We support per-core readings
+
+    // We do it this way to ensure correct nesting of reads
```

Comment:
I think we can add a bit more to this. Something like "The quickest, and most important, reading is krun_read_core_cycles, which is the innermost call before and after benchmarking. Although APERF/MPERF have separate functions, they are really one operation, so they are done in the same order before/after benchmarking (not, as you might expect, in reverse order after benchmarking)." Well, that's a bit rubbish, but you get the idea I hope!